velox/functions/prestosql/tests/MapFromEntriesTest.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace {
+class MapFromEntriesTest : public FunctionBaseTest {
```

Comment:
This function can be implemented in a zero-copy fashion as a vector function. We can reuse the input's offsets and sizes buffers as well as keys and values vectors to construct the result map. Vector function implementation will also allow us to support all types, not just primitive types for input. See velox/functions/prestosql/Map.cpp for an example code for checking for duplicate keys.

Reply:
Thanks for looking at my updated diff, @mbasmanova .

I understand your concern about lack of zero copy optimization for the generic-typed input. More general speaking, would you help me understand better the team criteria when a velox developer should use simple function api vs vector function api? I raised this question because you also suggest to reimplement `shuffle(array)` via vector function api.

Is supporting generic typed input/output + zero copy optimization the ONLY criteria?

Reply:
We recommend to use simple function API whenever possible. However, if simple function API cannot be used to implement all the necessary functionality or if vector function can be significantly faster, then vector function API should be used.